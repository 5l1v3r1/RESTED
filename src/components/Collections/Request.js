import React from 'react';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';
import { ListGroup } from 'react-bootstrap';
import flow from 'lodash.flow';

import IconButton from 'components/IconButton';
import * as Actions from 'store/collections/actions';

import { StyledRequest, AsideButtons, MainContent } from './StyledComponents';
import * as Type from './dropTypes';

/**
 * Specifies the drag source contract.
 * Only `beginDrag` function is required.
 */
const requestSource = {
  canDrag() {
    // TODO Disallow drag if editing name
    return true;
  },

  beginDrag({ id, index, collectionIndex }) {
    return { id, index, collectionIndex };
  },
};

const requestTarget = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dragCollectionIndex = monitor.getItem().collectionIndex;
    const dragUUID = monitor.getItem().id;
    const hoverIndex = props.index;
    const hoverCollectionIndex = props.collectionIndex;
    const hoverUUID = props.id;

    // Abort further processing if dragged item is hovering over itself
    if (dragUUID === hoverUUID) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().index = hoverIndex;
    monitor.getItem().collectionIndex = hoverCollectionIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderRequest({
      // Source
      collectionIndex: dragCollectionIndex,
      requestIndex: dragIndex,
    }, {
      // Target
      collectionIndex: hoverCollectionIndex,
      requestIndex: hoverIndex,
    });
  },
};

function Request(props) {
  const {
    connectDragSource,
    connectDropTarget,
    isDragging,
    id,
    collectionIndex,
    method,
    url,
    deleteRequest,
  } = props;

  return connectDragSource(connectDropTarget(
    <div> {/* Need a wrapper div for React DnD support */}
      <StyledRequest isDragging={isDragging}>
        <ListGroup componentClass="div">
          <div className="list-group-item">
            <AsideButtons>
              <IconButton
                tooltip="Edit"
                icon="cog"
              />
              <IconButton
                tooltip="Delete"
                icon="trash"
                onClick={() => deleteRequest(id, collectionIndex)}
              />
            </AsideButtons>

            <MainContent>
              <h4>{method}</h4>
              {url}
            </MainContent>
          </div>
        </ListGroup>
      </StyledRequest>
    </div>,
  ));
}

export default flow(
  DropTarget(Type.Request, requestTarget, connector => ({
    connectDropTarget: connector.dropTarget(),
  })),
  DragSource(Type.Request, requestSource, (connector, monitor) => ({
    connectDragSource: connector.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  connect(null, Actions),
)(Request);

