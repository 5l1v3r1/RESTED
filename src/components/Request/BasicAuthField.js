import React, { PropTypes } from 'react';
import { Col, FormGroup, FormControl } from 'react-bootstrap';

import Collapsable from '../Collapsable';

export function BasicAuthField(fields) {
  return (
    <Collapsable
      title="Basic auth"
      id="basicAuth"
    >
      <FormGroup controlId="method">
        <Col xs={5}>
          <FormControl
            type="text"
            placeholder="Username"
            {...fields.basicAuth.username.input}
          />
        </Col>
        <Col xs={5}>
          <FormControl
            type="text"
            placeholder="Password"
            {...fields.basicAuth.password.input}
          />
        </Col>
        <Col xs={2}>
          Show password?
        </Col>
      </FormGroup>
    </Collapsable>
  );
}

BasicAuthField.propTypes = {
  basicAuth: PropTypes.shape({
    username: PropTypes.shape({
      input: PropTypes.shape({}).isRequired,
    }).isRequired,
    password: PropTypes.shape({
      input: PropTypes.shape({}).isRequired,
    }).isRequired,
  }).isRequired,
};

export default BasicAuthField;

