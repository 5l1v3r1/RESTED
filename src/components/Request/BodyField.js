import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Field, FieldArray } from 'redux-form';
import { FormGroup, FormControl, Checkbox, Row, Col } from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsable from 'components/Collapsable';
import IconButton from 'components/IconButton';
import * as RequestActions from 'store/request/actions';

import { UnstyledButton } from './StyledComponents';

function renderField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function renderFormDataFields({ fields, meta, setUseFormData }) {
  return (
    <Collapsable
      title="Request body"
      id="requestBody"
    >
      <Col xs={12}>
        <Checkbox
          checked
          onChange={() => { setUseFormData(false); }}
        >
          Use form data
        </Checkbox>
      </Col>
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`requestBodyGroup${key}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={renderField}
              placeholder="Name"
            />
          </Col>
          <Col xs={6}>
            <Field
              name={`${field}.value`}
              component={renderField}
              placeholder="Value"
            />
          </Col>
          <Col xs={1}>
            <IconButton
              id={`removeFormDataButton${key}`}
              tooltip="Remove form field"
              icon="trash"
              onClick={() => fields.remove(key)}
            />
          </Col>
        </FormGroup>
      ))}
      <Row>
        <Col xs={12}>
          <UnstyledButton
            id="addParameter"
            bsStyle="link"
            onClick={() => fields.push({})}
          >
            <Fonticon icon="plus" />
            Add parameter
          </UnstyledButton>
        </Col>
      </Row>
    </Collapsable>
  );
}

/* eslint-disable react/no-unused-prop-types */
renderFormDataFields.propTypes = {
  setUseFormData: PropTypes.func.isRequired,
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

function renderDataField({ input, setUseFormData }) {
  return (
    <Collapsable
      title="Request body"
      id="requestBody"
    >
      <Col xs={12}>
        <Checkbox
          onChange={() => { setUseFormData(true); }}
        >
          Use form data
        </Checkbox>
      </Col>

      <Col xs={12}>
        <FormGroup controlId="requestBody">
          <FormControl
            componentClass="textarea"
            rows={10}
            {...input}
          />
        </FormGroup>
      </Col>
    </Collapsable>
  );
}

renderDataField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  setUseFormData: PropTypes.func.isRequired,
};

export function BodyField({ useFormData, setUseFormData }) {
  if (useFormData) {
    return (
      <FieldArray
        name="formData"
        component={renderFormDataFields}
        setUseFormData={setUseFormData}
      />
    );
  }

  return (
    <Field
      name="data"
      component={renderDataField}
      setUseFormData={setUseFormData}
    />
  );
}

BodyField.propTypes = {
  useFormData: PropTypes.bool,
  setUseFormData: PropTypes.func.isRequired,
};

const mapStateToProps = ({ request: { useFormData } }) => ({
  useFormData,
});

export default connect(mapStateToProps, RequestActions)(BodyField);

