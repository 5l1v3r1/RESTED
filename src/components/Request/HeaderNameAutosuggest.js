import React, { PropTypes } from 'react';
import classNames from 'classnames';
import Autosuggest from 'react-autosuggest';

import headers from 'constants/commonHeaders';
import { SuggestWrapper } from './StyledComponents';

const maxEntries = 7;

// Calculate suggestions for any given input value.
const getSuggestions = value => {
  const inputValue = value.trim().toLowerCase();
  const inputLength = inputValue.length;

  return inputLength === 0
    ? []
    : headers
      .filter(header =>
        header.toLowerCase().includes(inputValue),
      )
      .slice(0, maxEntries);
};

// Teach Autosuggest how to calculate the input value for
// every given suggestion.
const getSuggestionValue = suggestion => suggestion;

// Ensure the user does not send a request when selecting using enter
function preventDefaultOnEnter(event, { method }) {
  if (method === 'enter') {
    event.preventDefault();
  }
}

const renderSuggestion = suggestion => (
  <div>
    {suggestion}
  </div>
);

const renderInputComponent = ({ className, ...rest }) => (
  <input className={classNames(className, 'form-control')} {...rest} />
);

renderInputComponent.propTypes = {
  className: PropTypes.string,
};

export default class HeaderNameAutosuggest extends React.PureComponent {
  static propTypes = {
    input: PropTypes.shape({}).isRequired,
    placeholder: PropTypes.string.isRequired,
  };

  state = {
    suggestions: [],
  };

  onSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: getSuggestions(value),
    });
  };

  // When the suggest is closed or emptied
  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: [],
    });
  };

  render() {
    const { input, placeholder } = this.props;

    const inputProps = {
      ...input,
      placeholder,
      onChange: (_, { newValue }) => input.onChange(newValue),
    };

    return (
      <SuggestWrapper>
        <Autosuggest
          suggestions={this.state.suggestions}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          onSuggestionSelected={preventDefaultOnEnter}
          getSuggestionValue={getSuggestionValue}
          renderInputComponent={renderInputComponent}
          renderSuggestion={renderSuggestion}
          inputProps={inputProps}
        />
      </SuggestWrapper>
    );
  }
}

