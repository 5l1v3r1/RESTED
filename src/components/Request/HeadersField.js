import React, { PropTypes } from 'react';
import { Field } from 'redux-form';
import {
  FormGroup,
  FormControl,
  Button,
  Col,
} from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsable from 'components/Collapsable';
import IconButton from 'components/IconButton';

function renderField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function HeadersField({ meta, fields }) {
  return (
    <Collapsable
      title="Headers"
      id="headers"
    >
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`header.${field}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={renderField}
              placeholder="Name"
            />
          </Col>
          <Col xs={6}>
            <Field
              name={`${field}.value`}
              component={renderField}
              placeholder="Value"
            />
          </Col>
          <Col xs={1}>
            <IconButton
              id={`removeHeaderButton${key}`}
              tooltip="Remove header"
              icon="trash"
              onClick={() => fields.remove(key)}
            />
          </Col>
        </FormGroup>
      ))}

      <Col xs={12}>
        <Button
          id="addHeaderButton"
          onClick={() => fields.push({})}
        >
          <Fonticon icon="plus" />
          Add header
        </Button>
      </Col>
    </Collapsable>
  );
}

// ESLint is not smart enough to see that these are used
/* eslint-disable react/no-unused-prop-types */
HeadersField.propTypes = {
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

export default HeadersField;

