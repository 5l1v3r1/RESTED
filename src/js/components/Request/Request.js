import React from 'react';
import { Col } from 'react-bootstrap';

export default function Request() {
  return (
    <Col xs={12} sm={8}>
      <h2>
        Request
      </h2>
    </Col>
  );
}

