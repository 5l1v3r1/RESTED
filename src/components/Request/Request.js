import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel } from 'react-bootstrap';

import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

import * as Actions from '../../store/request/actions';
import { DEFAULT_REQUEST } from '../../constants/constants';

function Titlebar() {
  return (
    <h2>
      Request
    </h2>
  );
}

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <form onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: 'request',
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

// TODO this
/* eslint-disable no-func-assign */
Request = reduxForm(formOptions)(Request);
Request = connect(mapStateToProps, Actions)(Request);
/* eslint-enable no-func-assign */
export default Request;

