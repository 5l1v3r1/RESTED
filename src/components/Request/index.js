import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel, Form } from 'react-bootstrap';
import flow from 'lodash.flow';

import * as requestActions from 'store/request/actions';
import requestValidation from 'utils/requestValidation';
import { DEFAULT_REQUEST } from 'constants/constants';

import Titlebar from './Titlebar';
import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

export const requestForm = 'request';

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <Form horizontal onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </Form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: requestForm,
  validate: requestValidation,
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

export default flow(
  reduxForm(formOptions),
  connect(mapStateToProps, requestActions),
)(Request);

