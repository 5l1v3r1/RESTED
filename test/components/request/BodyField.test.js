import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider, connect } from 'react-redux';
import { reduxForm } from 'redux-form';

/* eslint-disable import/no-unresolved */
import { BodyField } from 'components/Request/BodyField';
import makeStore from '../../makeStore';

describe('BodyField', () => {
  const store = makeStore({});

  it('should match the previous snapshot when !useFormData', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData={false}
        setUseFormData={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match the previous snapshot when !!useFormData', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData
        setUseFormData={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render the formData fields', () => {
    const initialValues = {
      formData: [
        {
          name: 'foo',
          value: 'something',
        }, {
          name: 'bar',
          value: 'something2',
        },
      ],
    };

    let Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData
        setUseFormData={() => {}}
      />
    ));

    Decorated = connect(() => ({
      initialValues,
    }))(Decorated);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>
    );

    // Filter out the initial checkbox
    const inputs = tree
      .find('input')
      .filterWhere(i => i.prop('type') !== 'checkbox');

    expect(inputs.length).toBe(4);

    let i = 0;
    initialValues.formData.forEach(header => {
      const nameField = inputs.at(i++); // eslint-disable-line no-plusplus
      const valField = inputs.at(i++);  // eslint-disable-line no-plusplus

      expect(nameField.prop('placeholder')).toBe('Name');
      expect(valField.prop('placeholder')).toBe('Value');
      expect(nameField.prop('value')).toBe(header.name);
      expect(valField.prop('value')).toBe(header.value);
    });
  });

  it('should render the data textarea', () => {
    const initialValues = {
      data: 'foo crux bluxorb',
    };

    let Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData={false}
        setUseFormData={() => {}}
      />
    ));

    Decorated = connect(() => ({
      initialValues,
    }))(Decorated);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>
    );

    const input = tree.find('textarea');

    expect(input.length).toBe(1);
  });

  it('should call setUseFormData when !!useFormData checkbox is clicked', () => {
    const setUseFormData = jest.fn();

    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData
        setUseFormData={setUseFormData}
      />
    ));

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>
    );

    expect(setUseFormData).not.toHaveBeenCalled();

    const checkbox = tree.find('input').first();
    checkbox.simulate('change');

    expect(setUseFormData).toHaveBeenCalledWith(false);
  });

  it('should call setUseFormData when !useFormData checkbox is clicked', () => {
    const setUseFormData = jest.fn();

    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        useFormData={false}
        setUseFormData={setUseFormData}
      />
    ));

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>
    );

    expect(setUseFormData).not.toHaveBeenCalled();

    const checkbox = tree.find('input').first();
    checkbox.simulate('change');

    expect(setUseFormData).toHaveBeenCalledWith(true);
  });
});

