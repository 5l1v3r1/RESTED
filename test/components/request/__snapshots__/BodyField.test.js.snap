// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyField should match the previous snapshot when !!useFormData 1`] = `
<div
  className="row"
>
  <button
    className="btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4>
      Request body
      <i
        className="fa fa-angle-right"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse col-xs-12"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="checkbox"
        style={undefined}
      >
        <label>
          <input
            checked={true}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          Use form data
        </label>
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <button
        className="btn btn-default"
        disabled={false}
        id="addParameter"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa fa-plus"
          role="presentation"
        />
        Add parameter
      </button>
    </div>
  </div>
</div>
`;

exports[`BodyField should match the previous snapshot when !useFormData 1`] = `
<div
  className="row"
>
  <button
    className="btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4>
      Request body
      <i
        className="fa fa-angle-right"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse col-xs-12"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="checkbox"
        style={undefined}
      >
        <label>
          <input
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          Use form data
        </label>
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control"
          id="requestBody"
          name="data"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          rows={10}
          type={undefined}
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
